### ISSUE TITLE
no such attribute 'netrc' in 'http_file' rule

### ISSUE THREAD

**Issue by hrsgn** [apache/incubator-heron#3600]
**Describe the bug**
when i use bazel(bazel build --config=centos heron/...) to buid the heron source code from "https://github.com/apache/incubator-heron.git",there is some error as below:
Loading: 
Loading: 0 packages loaded
ERROR: /home/zw/.cache/bazel/_bazel_zw/0d94d7c020c06b4f73081720ff5e8130/external/maven/defs.bzl:3:5: //external:ant_ant_1_6_5: no such attribute 'netrc' in 'http_file' rule
ERROR: /home/zw/.cache/bazel/_bazel_zw/0d94d7c020c06b4f73081720ff5e8130/external/maven/defs.bzl:9:5: //external:ant_ant_jar_sources_1_6_5: no such attribute 'netrc' in 'http_file' rule
ERROR: /home/zw/.cache/bazel/_bazel_zw/0d94d7c020c06b4f73081720ff5e8130/external/maven/defs.bzl:15:5: //external:antlr_antlr_2_7_7: no such attribute 'netrc' in 'http_file' rule
ERROR: /home/zw/.cache/bazel/_bazel_zw/0d94d7c020c06b4f73081720ff5e8130/external/maven/defs.bzl:21:5: //external:aopalliance_aopalliance_1_0: no such attribute 'netrc' in 'http_file' rule
ERROR: /home/zw/.cache/bazel/_bazel_zw/0d94d7c020c06b4f73081720ff5e8130/external/maven/defs.bzl:27:5: //external:aopalliance_aopalliance_jar_sources_1_0: no such attribute 'netrc' in 'http_file' rule
ERROR: /home/zw/.cache/bazel/_bazel_zw/0d94d7c020c06b4f73081720ff5e8130/external/maven/defs.bzl:33:5: //external:cglib_cglib_3_1: no such attribute 'netrc' in 'http_file' rule
ERROR: /home/zw/.cache/bazel/_bazel_zw/0d94d7c020c06b4f73081720ff5e8130/external/maven/defs.bzl:39:5: //external:cglib_cglib_jar_sources_3_1: no such attribute 'netrc' in 'http_file' rule
ERROR: /home/zw/.cache/bazel/_bazel_zw/0d94d7c020c06b4f73081720ff5e8130/external/maven/defs.bzl:45:5: //external:classworlds_classworlds_1_1_alpha_2: no such attribute 'netrc' in 'http_file' rule
......
ERROR: /home/zw/.cache/bazel/_bazel_zw/0d94d7c020c06b4f73081720ff5e8130/external/maven/defs.bzl:2697:5: //external:xmlenc_xmlenc_0_52: no such attribute 'netrc' in 'http_file' rule
ERROR: error loading package '': Encountered error while reading extension file 'python/repositories.bzl': no such package '@rules_python//python': error loading package 'external': Could not load //external package
ERROR: error loading package '': Encountered error while reading extension file 'python/repositories.bzl': no such package '@rules_python//python': error loading package 'external': Could not load //external package
INFO: Elapsed time: 0.788s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (0 packages loaded)
FAILED: Build did NOT complete successfully (0 packages loaded)

**To Reproduce**
1.$ git clone https://github.com/apache/incubator-heron.git && cd incubator-heron
   $ ./bazel_configure.py
2. cd ~
   wget -O /tmp/bazel.sh https://github.com/bazelbuild/bazel/releases/download/0.26.0/bazel-0.26.0-installer-linux-x86_64.sh
   chmod +x /tmp/bazel.sh
 /tmp/bazel.sh --user
3. cd   incubator-heron 
 bazel build --config=centos heron/...

**Operating System**
 - OS: centOS7
 - python3.8.5

**Additional context**

[zw@centos01 incubator-heron]$ cd  /home/zw/.cache/bazel/_bazel_zw/0d94d7c020c06b4f73081720ff5e8130/external/maven/
[zw@centos01 maven]$ ll
总用量 508
-rw-rw-r--. 1 zw zw   272452  8月   5 17:41 BUILD
-rw-rw-r--. 1 zw zw    1063     8月   5 17:41 compat_repository.bzl
-rw-rw-r--. 1 zw zw  234652   8月   5 17:41 defs.bzl
lrwxrwxrwx. 1 zw zw     43      8月   5 17:40 imported_maven_install.json -> /home/zw/incubator-heron/maven_install.json
-rw-rw-r--. 1 zw zw       0        8月   5 17:41 netrc
-rw-rw-r--. 1 zw zw     111      8月   5 17:41 WORKSPACE
[zw@centos01 maven]$ more defs.bzl 
load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_file")
def pinned_maven_install():
    http_file(
        name = "ant_ant_1_6_5",
        sha256 = "f06a601c718a7c9262d74b7ec3baad14c82584e89235089b4f821d6a44d9e1e4",
        netrc = "../maven/netrc",
        urls = ["https://jcenter.bintray.com/ant/ant/1.6.5/ant-1.6.5.jar", "https://maven.google.com/ant/ant/1.6.5/ant-1.6.5.jar", "ht
tps://repo1.maven.org/maven2/ant/ant/1.6.5/ant-1.6.5.jar"],
    )
--More--(0%)

**Comment by joshfischer1108** [comment]
Hi,

It looks like  you are using Bazel version 0.26.0.  At this time Heron needs to be built with Bazel 3.4.1. 

I found where the documentation is incorrect.  We will get this fixed ASAP.

**Comment by joshfischer1108** [comment]
I've checked the documentation.  

If you are building off of master  you need to reference these docs: https://heron.incubator.apache.org/docs/next/getting-started-local-single-node

If you are building off of the `0.20.2-incubating` release you need to follow these: https://heron.incubator.apache.org/docs/compiling-linux

Let me know if you find any other issues.  We could add some work to make the website a little more clear.  Thank you for filing the issue.

**Comment by hrsgn** [comment]
> Hi,
> 
> It looks like you are using Bazel version 0.26.0. At this time Heron needs to be built with Bazel 3.4.1.
> 
> I found where the documentation is incorrect. We will get this fixed ASAP.

Thank you very much, and there is a new problem as below:

[zw@centos01 incubator-heron]$ bazel build --config=centos heron/...
DEBUG: /home/zw/.cache/bazel/_bazel_zw/0d94d7c020c06b4f73081720ff5e8130/external/bazel_tools/tools/cpp/lib_cc_configure.bzl:118:10: 
Auto-Configuration Warning: CC with -fuse-ld=gold returned 0, but its -v output didn't contain 'gold', falling back to the default linker.
INFO: SHA256 (https://archive.apache.org/dist/zookeeper/zookeeper-3.5.8/apache-zookeeper-3.5.8.tar.gz) = bafc0abe7da696a2020ba11b8ce7d06f6e28e9bf1e5504de09be25b8b589777d
DEBUG: Rule 'org_apache_zookeeper' indicated that a canonical reproducible form can be obtained by modifying arguments sha256 = "bafc0abe7da696a2020ba11b8ce7d06f6e28e9bf1e5504de09be25b8b589777d"
DEBUG: Repository org_apache_zookeeper instantiated at:
  no stack (--record_rule_instantiation_callstack not enabled)
Repository rule http_archive defined at:
  /home/zw/.cache/bazel/_bazel_zw/0d94d7c020c06b4f73081720ff5e8130/external/bazel_tools/tools/build_defs/repo/http.bzl:336:31: in <toplevel>
INFO: Repository io_bazel_rules_scala_scalatest instantiated at:
  no stack (--record_rule_instantiation_callstack not enabled)
Repository rule jvm_import_external defined at:
  /home/zw/.cache/bazel/_bazel_zw/0d94d7c020c06b4f73081720ff5e8130/external/io_bazel_rules_scala/scala/scala_maven_import_external.bzl:207:38: in <toplevel>
WARNING: Download from https://repo.maven.apache.org/maven2/org/scalatest/scalatest_2.12/3.0.5/scalatest_2.12-3.0.5.jar failed: class com.google.devtools.build.lib.bazel.repository.downloader.UnrecoverableHttpException Checksum was 978e7c4cb797f5d17e4cd8b928eda0c56d98d786537ed0917c409268888b9921 but wanted b416b5bcef6720da469a8d8a5726e457fc2d1cd5d316e1bc283aa75a2ae005e5
ERROR: An error occurred during the fetch of repository 'io_bazel_rules_scala_scalatest':
   java.io.IOException: Error downloading [https://repo.maven.apache.org/maven2/org/scalatest/scalatest_2.12/3.0.5/scalatest_2.12-3.0.5.jar] to /home/zw/.cache/bazel/_bazel_zw/0d94d7c020c06b4f73081720ff5e8130/external/io_bazel_rules_scala_scalatest/scalatest_2.12-3.0.5.jar: Checksum was 978e7c4cb797f5d17e4cd8b928eda0c56d98d786537ed0917c409268888b9921 but wanted b416b5bcef6720da469a8d8a5726e457fc2d1cd5d316e1bc283aa75a2ae005e5
**ERROR**: /home/zw/.cache/bazel/_bazel_zw/0d94d7c020c06b4f73081720ff5e8130/external/io_bazel_rules_scala/scala/scalatest/BUILD:5:13: @io_bazel_rules_scala//scala/scalatest:scalatest depends on @io_bazel_rules_scala_scalatest//:io_bazel_rules_scala_scalatest in repository @io_bazel_rules_scala_scalatest which failed to fetch. no such package '@io_bazel_rules_scala_scalatest//': java.io.IOException: Error downloading [https://repo.maven.apache.org/maven2/org/scalatest/scalatest_2.12/3.0.5/scalatest_2.12-3.0.5.jar] to /home/zw/.cache/bazel/_bazel_zw/0d94d7c020c06b4f73081720ff5e8130/external/io_bazel_rules_scala_scalatest/scalatest_2.12-3.0.5.jar: Checksum was 978e7c4cb797f5d17e4cd8b928eda0c56d98d786537ed0917c409268888b9921 but wanted b416b5bcef6720da469a8d8a5726e457fc2d1cd5d316e1bc283aa75a2ae005e5
ERROR: Analysis of target '//heron/api/tests/scala:api-scala-test' failed; build aborted: Analysis failed
INFO: Elapsed time: 2757.191s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (418 packages loaded, 5422 targets configured)

**Comment by hrsgn** [comment]
> I've checked the documentation.
> 
> If you are building off of master you need to reference these docs: https://heron.incubator.apache.org/docs/next/getting-started-local-single-node
> 
> If you are building off of the `0.20.2-incubating` release you need to follow these: https://heron.incubator.apache.org/docs/compiling-linux
> 
> Let me know if you find any other issues. We could add some work to make the website a little more clear. Thank you for filing the issue.
Thank you very much. Another problem I encountered is that the python version required to build heron does not seem to be higher than 3.8.0

**Comment by hrsgn** [comment]
> I've checked the documentation.
> 
> If you are building off of master you need to reference these docs: https://heron.incubator.apache.org/docs/next/getting-started-local-single-node
> 
> If you are building off of the `0.20.2-incubating` release you need to follow these: https://heron.incubator.apache.org/docs/compiling-linux
> 
> Let me know if you find any other issues. We could add some work to make the website a little more clear. Thank you for filing the issue.

Thank you very much. Another problem I encountered is that the python version required to build heron does not seem to be higher than 3.8.0

**Comment by joshfischer1108** [comment]
I believe Heron is compatible with Python 3.6.x at this time.  Maybe @Code0x58 or @nicknezis could add a few more details.

**Comment by hrsgn** [comment]
> I believe Heron is compatible with Python 3.6.x at this time. Maybe @Code0x58 or @nicknezis could add a few more details.

[zw@centos01 incubator-heron]$ ./bazel_configure.py 
Platform Linux
Using C compiler          :     /usr/bin/gcc (4.8.5)
Using C++ compiler        :     /usr/bin/g++ (4.8.5)
Using C preprocessor      :     /usr/bin/cpp (4.8.5)
Using C++ preprocessor    :     /usr/bin/cpp (4.8.5)
Using linker              :     /usr/bin/ld.bfd
Using JDK                 :     /usr/local/java/jdk1.8.0_151
Using Automake            :     /usr/bin/automake (1.13.4)
Using Autoconf            :     /usr/bin/autoconf (2.69)
Using Make                :     /usr/bin/make (3.82)
Using Python3             :     **/usr/local/python3.8/bin/python3.8 (3.8.0)**
Using Libtool             :     /usr/bin/libtool (2.4.2)
Using archiver            :     /usr/bin/ar
Using coverage tool       :     /usr/bin/gcov
Using ant                 :     /usr/local/ant/apache-ant-1.10.8/bin/ant
Using dwp                 :     /usr/bin/dwp
Using nm                  :     /usr/bin/nm
Using objcopy             :     /usr/bin/objcopy
Using objdump             :     /usr/bin/objdump
Using strip               :     /usr/bin/strip
[zw@centos01 incubator-heron]$  bazel build --config=centos heron/...
WARNING: /home/zw/incubator-heron/heron/healthmgr/tests/java/BUILD:52:13: in srcs attribute of java_library rule //heron/healthmgr/tests/java:healthmgr-tests: please do not import '//heron/healthmgr/src/java:org/apache/heron/healthmgr/HealthManager.java' directly. You should either move the file to this package or depend on an appropriate rule there. Since this rule was created by the macro 'java_library', the error might have been caused by the macro implementation
INFO: Analyzed 612 targets (0 packages loaded, 0 targets configured).
INFO: Found 612 targets...
**ERROR: /home/zw/incubator-heron/third_party/python/cpplint/BUILD:5:11: PexPython third_party/python/cpplint/cpplint.pex failed (Exit 127)**
**/usr/bin/env: python3.8: No such file or directory**
INFO: Elapsed time: 2.569s, Critical Path: 0.05s
INFO: 0 processes.
FAILED: Build did NOT complete successfully

**Comment by hrsgn** [comment]
I switched to python3.6.11, the same error as before
**ERROR**: /home/zw/incubator-heron/third_party/python/cpplint/BUILD:5:11: PexPython third_party/python/cpplint/cpplint.pex failed (Exit 127)
/usr/bin/env: python3.6: No such file or directory

**Comment by Code0x58** [comment]
_**TL;DR:** your python3 binary is probably called `python` or `python3`, but PEX by default needs `python<major>.<minor>`_

one thing with PEX is that it tries to be explicit about what version of python it is for by default (you could argue it's good for reproducibility, but a pain for flexibility), so if you run code on a python3.6 interpreter, it will automatically make it so the generated PEX starts with `#!/usr/bin/env python3.6`, rather than more generic shebang lines calling `python3` or `python`.

It seems you don't have a `python3.6` file on your system `$PATH`, which you'll need to stop that _"No such file or directory"_. Centos may have a package or option to install the more explicit `pythonX.Y` link, or you may have to symlink or hardlink it yourself.

p.s. yep, 3.6 is what is tested in CI and the minimum compatible version, but there's a decent chance that later versions also work - I can confirm that the build and tests work with python3.8 as that is what ends up used if you run [`./vagrant/local-ci.sh`](https://github.com/apache/incubator-heron/blob/master/vagrant/local-ci.sh) (which does the same process as Travis CI).

**Comment by hrsgn** [comment]
> _**TL;DR:** your python3 binary is probably called `python` or `python3`, but PEX by default needs `python<major>.<minor>` where the major and minor version are the same as those used to build the PEX_
> 
> _**edit:** humm, your `./baze_configure.py` dump shows you do have a `python<major>.<minor>` - for some reason that isn't visible in the environment the PEX is being executed - is it a separate environment?_
> 
> one thing with PEX is that it tries to be explicit about what version of python it is for by default (you could argue it's good for reproducibility, but a pain for flexibility), so if you run code on a python3.6 interpreter, it will automatically make it so the generated PEX starts with `#!/usr/bin/env python3.6`, rather than more generic shebang lines calling `python3` or `python`.
> 
> It seems you don't have a `python3.6` file on your system `$PATH`, which you'll need to stop that _"No such file or directory"_. Centos may have a package or option to install the more explicit `pythonX.Y` link, or you may have to symlink or hardlink it yourself.
> 
> p.s. yep, 3.6 is what is tested in CI and the minimum compatible version, but there's a decent chance that later versions also work - I can confirm that the build and tests work with python3.8 as that is what ends up used if you run [`./vagrant/local-ci.sh`](https://github.com/apache/incubator-heron/blob/master/vagrant/local-ci.sh) (which does the same process as Travis CI).


@Code0x58 I’m sorry, I still haven’t found the reason. I haven’t used python before and don’t know much about python.

[zw@centos01 incubator-heron]$ **bazel build --config=centos heron/...**
WARNING: /home/zw/incubator-heron/heron/healthmgr/tests/java/BUILD:52:13: in srcs attribute of java_library rule //heron/healthmgr/tests/java:healthmgr-tests: please do not import '//heron/healthmgr/src/java:org/apache/heron/healthmgr/HealthManager.java' directly. You should either move the file to this package or depend on an appropriate rule there. Since this rule was created by the macro 'java_library', the error might have been caused by the macro implementation
INFO: Analyzed 612 targets (0 packages loaded, 0 targets configured).
INFO: Found 612 targets...
**ERROR: /home/zw/incubator-heron/third_party/python/cpplint/BUILD:5:11: PexPython third_party/python/cpplint/cpplint.pex failed (Exit 127)
/usr/bin/env: python3.6: No such file or directory**
INFO: Elapsed time: 1.696s, Critical Path: 0.04s
INFO: 0 processes.
FAILED: Build did NOT complete successfully

[zw@centos01 incubator-heron]$ **ls -l /usr/bin/python***
lrwxrwxrwx. 1 root root   32 8月   7 16:47 /usr/bin/python -> /usr/local/python3.6/bin/python3
lrwxrwxrwx. 1 root root    9 8月   5 00:27 /usr/bin/python2 -> python2.7
-rwxr-xr-x. 1 root root 7144 4月   2 21:17 /usr/bin/python2.7
lrwxrwxrwx. 1 root root   32 8月  10 00:41 /usr/bin/python3 -> /usr/local/python3.6/bin/python3

[zw@centos01 bin]$ **echo $PATH**
/usr/local/ant/apache-ant-1.10.8/bin:**/usr/local/python3.6/bin**:/usr/local/java/jdk1.8.0_151/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/home/zw/.local/bin:/home/zw/bin
[zw@centos01 incubator-heron]$ **cd /usr/local/python3.6/bin/ && ll**
总用量 24864
lrwxrwxrwx. 1 root root        8 8月   7 16:34 2to3 -> 2to3-3.6
-rwxr-xr-x. 1 root root      111 8月   7 16:34 2to3-3.6
-rwxr-xr-x. 1 root root      252 8月   7 16:34 easy_install-3.6
lrwxrwxrwx. 1 root root        7 8月   7 16:34 idle3 -> idle3.6
-rwxr-xr-x. 1 root root      109 8月   7 16:34 idle3.6
-rwxr-xr-x. 1 root root      234 8月   7 16:34 pip3
-rwxr-xr-x. 1 root root      234 8月   7 16:34 pip3.6
lrwxrwxrwx. 1 root root        8 8月   7 16:34 pydoc3 -> pydoc3.6
-rwxr-xr-x. 1 root root       94 8月   7 16:34 pydoc3.6
lrwxrwxrwx. 1 root root        9 8月   7 16:34 python3 -> python3.6
-rwxr-xr-x. 2 root root 12710008 8月   7 16:33 python3.6
lrwxrwxrwx. 1 root root       17 8月   7 16:34 python3.6-config -> python3.6m-config
-rwxr-xr-x. 2 root root 12710008 8月   7 16:33 python3.6m
-rwxr-xr-x. 1 root root     3107 8月   7 16:34 python3.6m-config
lrwxrwxrwx. 1 root root       16 8月   7 16:34 python3-config -> python3.6-config
lrwxrwxrwx. 1 root root       10 8月   7 16:34 pyvenv -> pyvenv-3.6
-rwxr-xr-x. 1 root root      451 8月   7 16:34 pyvenv-3.6

[zw@centos01 bin]$ **python3**
Python 3.6.11 (default, Aug  7 2020, 16:32:35) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-39)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> 

[zw@centos01 bin]$ **more**  /home/zw/incubator-heron/third_party/python/cpplint/**cpplint.py** 
#!/usr/bin/env python3
#
# Copyright (c) 2009 Google Inc. All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
...

**Comment by Code0x58** [comment]
Humm, that shell session makes it look like it should work :s Just to double check, could you run this code and share the output?
```bash
# see if the python interpreters resolve, and see what they resolve to
for label in python3 python3.6; do
    path="$(command -v $label)"
    echo "for $label:"
    echo "    path: $path"
    echo "    path breakdown:"
    namei "$path"
    echo "    version: $($label -V)"
done
# this is the file that is generated and executed, lets see the shebang line
head -n1 ./bazel-bin/third_party/python/cpplint/cpplint.pex
# try running the binary, just to check it's not the bazel environment that is different
./bazel-bin/third_party/python/cpplint/cpplint.pex --help | head -n5
```

**Comment by hrsgn** [comment]
[zw@centos01 incubator-heron]$ more test1.py 
```
#!/usr/bin/env python3
# see if the python interpreters resolve, and see what they resolve to
for label in python3 python3.6; do
    path="$(command -v $label)"
    echo "for $label:"
    echo "    path: $path"
    echo "    path breakdown:"
    namei "$path"
    echo "    version: $($label -V)"
done
# this is the file that is generated and executed, lets see the shebang line
head -n1 ./bazel-bin/third_party/python/cpplint/cpplint.pex
# try running the binary, just to check it's not the bazel environment that is different
./bazel-bin/third_party/python/cpplint/cpplint.pex --help | head -n5
```
[zw@centos01 incubator-heron]$ ./test1.py 
  File "./test1.py", line 3
    for label in python3 python3.6; do
                               ^
SyntaxError: invalid syntax

@Code0x58

**Comment by hrsgn** [comment]
[zw@centos01 incubator-heron]$ ./test.sh
```
for python3:
    path: /usr/local/python3.6/bin/python3
    path breakdown:
f: /usr/local/python3.6/bin/python3
 d /
 d usr
 d local
 d python3.6
 d bin
 l python3 -> python3.6
   - python3.6
    version: Python 3.6.11
for python3.6:
    path: /usr/local/python3.6/bin/python3.6
    path breakdown:
f: /usr/local/python3.6/bin/python3.6
 d /
 d usr
 d local
 d python3.6
 d bin
 - python3.6
    version: Python 3.6.11
head: 无法打开"./bazel-bin/third_party/python/cpplint/cpplint.pex" 读取数据: 没有那个文件或目录
./test.sh:行13: ./bazel-bin/third_party/python/cpplint/cpplint.pex: 没有那个文件或目录
```
[zw@centos01 incubator-heron]$ cd ./bazel-bin/ && ll
```
总用量 0
drwxrwxr-x. 2 zw zw  28 8月  10 00:09 config
drwxrwxr-x. 9 zw zw 204 8月   6 22:39 external
drwxrwxr-x. 5 zw zw  49 8月   6 22:51 heron
```

**Comment by Code0x58** [comment]
the `command -v` and `namei` parts confirm your python works in your environment, so that's a :+1: 

The last two lines of the script didn't find the generated `cpplint.pex` - have you done a `bazel clean` since your last `bazel build --config=centos heron/...`?

I wonder if the _"/usr/bin/env: python3.6: No such file or directory"_ is caused by your `$PATH` not having been exported, so when bazel runs it does not get the value. Try running `export PATH` in your shell before you call bazel build. You can see all exported variables with `export -p`

**Comment by hrsgn** [comment]
@Code0x58 The problem still exists

[zw@centos01 incubator-heron]$ export -p
```
declare -x ANT_HOME="/usr/local/ant/apache-ant-1.10.8"
declare -x CLASSPATH=".:/usr/local/java/jdk1.8.0_151/lib/dt.jar:/usr/local/java/jdk1.8.0_151/lib/tools.jar"
declare -x HISTCONTROL="ignoredups"
declare -x HISTSIZE="1000"
declare -x HOME="/home/zw"
declare -x HOSTNAME="centos01"
declare -x JAVA_HOME="/usr/local/java/jdk1.8.0_151"
declare -x JRE_HOME="/usr/local/java/jdk1.8.0_151/jre"
declare -x LANG="zh_CN.UTF-8"
declare -x LESSOPEN="||/usr/bin/lesspipe.sh %s"
declare -x LOGNAME="zw"
declare -x LS_COLORS="rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=01;05;37;41:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.tzo=01;31:*.t7z=01;31:*.zip=01;31:*.z=01;31:*.Z=01;31:*.dz=01;31:*.gz=01;31:*.lrz=01;31:*.lz=01;31:*.lzo=01;31:*.xz=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.alz=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.cab=01;31:*.jpg=01;35:*.jpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.axv=01;35:*.anx=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=01;36:*.au=01;36:*.flac=01;36:*.mid=01;36:*.midi=01;36:*.mka=01;36:*.mp3=01;36:*.mpc=01;36:*.ogg=01;36:*.ra=01;36:*.wav=01;36:*.axa=01;36:*.oga=01;36:*.spx=01;36:*.xspf=01;36:"
declare -x MAIL="/var/spool/mail/zw"
declare -x OLDPWD="/usr/local"
declare -x PATH="/usr/local/ant/apache-ant-1.10.8/bin:/usr/local/python3.6/bin:/usr/local/java/jdk1.8.0_151/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/home/zw/.local/bin:/home/zw/bin"
declare -x PWD="/home/zw/incubator-heron"
declare -x SELINUX_LEVEL_REQUESTED=""
declare -x SELINUX_ROLE_REQUESTED=""
declare -x SELINUX_USE_CURRENT_RANGE=""
declare -x SHELL="/bin/bash"
declare -x SHLVL="1"
declare -x SSH_CLIENT="192.168.36.1 3973 22"
declare -x SSH_CONNECTION="192.168.36.1 3973 192.168.36.200 22"
declare -x SSH_TTY="/dev/pts/0"
declare -x TERM="vt100"
declare -x USER="zw"
declare -x XDG_RUNTIME_DIR="/run/user/1000"
declare -x XDG_SESSION_ID="1"
```

**Comment by joshfischer1108** [comment]
@hrsgn has this issue been resolved?

**Comment by hrsgn** [comment]
@joshfischer1108 

sorry, not yet, I have been preparing for an exam in recent months, I will try to solve this problem after the exam is over

**Comment by nicknezis** [comment]
What version of Bazel are you using? This error sounds like a similar issue I experienced when using an older Bazel.

**Comment by hrsgn** [comment]
> Hi,
> 
> It looks like you are using Bazel version 0.26.0. At this time Heron needs to be built with Bazel 3.4.1.
> 
> I found where the documentation is incorrect. We will get this fixed ASAP.

@nicknezis I have changed to Bazel 3.4.1

**Comment by joshfischer1108** [comment]
One the the harder to  manage parts of Bazel is that it often has breaking changes between versions.  Nick recently updated the codebase to be built with Bazel 3.7.0

If you are ever unsure of what version is required you can check the `.bazelversion` file [here](https://github.com/apache/incubator-heron/blob/master/.bazelversion).  The placement of this file in the repository should also give you a warning if you use a different version of Bazel in the project directory.

It would look something like this depending on your machine.
```
ERROR: The project you're trying to build requires Bazel 3.7.0 (specified in /path/to/incubator-heron/.bazelversion), but it wasn't found in /$HOME/.bazel/bin.

Bazel binaries for all official releases can be downloaded from here:
  https://github.com/bazelbuild/bazel/releases

You can download the required version directly using this command:
  (cd "/$HOME/.bazel/bin" && curl -LO https://releases.bazel.build/3.7.0/release/bazel-3.7.0-darwin-x86_64 && chmod +x bazel-3.7.0-darwin-x86_64)
```

**Comment by hrsgn** [comment]
I reinstalled heron and bazel (3.7.0), but still have the following problems. When I am free, I will reinstall all dependencies in a new environment.

> > [zw@centos01 incubator-heron]$ bazel build --config=centos heron/...
DEBUG: /home/zw/.cache/bazel/_bazel_zw/0d94d7c020c06b4f73081720ff5e8130/external/bazel_tools/tools/cpp/lib_cc_configure.bzl:118:10: 
Auto-Configuration Warning: CC with -fuse-ld=gold returned 0, but its -v output didn't contain 'gold', falling back to the default linker.
INFO: SHA256 (https://archive.apache.org/dist/zookeeper/zookeeper-3.5.8/apache-zookeeper-3.5.8.tar.gz) = bafc0abe7da696a2020ba11b8ce7d06f6e28e9bf1e5504de09be25b8b589777d
DEBUG: Rule 'org_apache_zookeeper' indicated that a canonical reproducible form can be obtained by modifying arguments sha256 = "bafc0abe7da696a2020ba11b8ce7d06f6e28e9bf1e5504de09be25b8b589777d"
DEBUG: Repository org_apache_zookeeper instantiated at:
  /home/zw/incubator-heron/WORKSPACE:263:13: in <toplevel>
Repository rule http_archive defined at:
  /home/zw/.cache/bazel/_bazel_zw/0d94d7c020c06b4f73081720ff5e8130/external/bazel_tools/tools/build_defs/repo/http.bzl:336:31: in <toplevel>
WARNING: /home/zw/incubator-heron/heron/healthmgr/tests/java/BUILD:52:13: in srcs attribute of java_library rule //heron/healthmgr/tests/java:healthmgr-tests: please do not import '//heron/healthmgr/src/java:org/apache/heron/healthmgr/HealthManager.java' directly. You should either move the file to this package or depend on an appropriate rule there. Since this rule was created by the macro 'java_library', the error might have been caused by the macro implementation
INFO: Analyzed 610 targets (606 packages loaded, 6340 targets configured).
INFO: Found 610 targets...
ERROR: /home/zw/incubator-heron/heron/shell/src/python/BUILD:17:11: PexPython heron/shell/src/python/heron-shell.pex failed (Exit 127): pex_wrapper.pex failed: error executing command bazel-out/host/bin/tools/rules/pex/pex_wrapper.pex --not-zip-safe --pex-root .pex --entry-point heron.shell.src.python.main --output-file ... (remaining 6 argument(s) skipped)
/usr/bin/env: python3.6: No such file or directory
INFO: Elapsed time: 7090.379s, Critical Path: 55.66s
INFO: 444 processes: 29 internal, 415 local.
FAILED: Build did NOT complete successfully

**Comment by nicknezis** [comment]
@hrsgn Were you able to resolve your issue? I'm closing the issue for now, but feel free to post another issue if you still are having environment problems. You can also join us in the Heron Slack to discuss.

**Comment by idlecloud** [comment]
I hava the same question, I use python 3.6.5, bazel 4.1.0
